ouroboros-network/src/Ouroboros/Network/PeerSelection/Churn.hs

```diff
@@ -44,30 +59,52 @@ peerChurnGovernor :: forall m peeraddr.
                   -> StrictTVar m ChurnMode
                   -> StdGen
                   -> STM m FetchMode
-                  -> PeerSelectionTargets
+                  -> TargetsSelector
                   -> StrictTVar m PeerSelectionTargets
                   -> STM m UseBootstrapPeers
+                  -> StrictTMVar m LedgerStateJudgement
+                  -- ^ Peer selection govnr chooses the ledger state that churn
+                  -- govnr should use for setting targets to keep everything in sync
+                  -> Bool
                   -> m Void
 peerChurnGovernor tracer deadlineChurnInterval bulkChurnInterval psOverallTimeout
-                  _metrics churnModeVar inRng getFetchMode base peerSelectionVar
-                  getUseBootstrapPeers = do
+                  _metrics churnModeVar inRng getFetchMode targetsSelector peerSelectionVar
+                  getUseBootstrapPeers mutexPeerSelection useGenesis = do
   -- Wait a while so that not only the closest peers have had the time
   -- to become warm.
   startTs0 <- getMonotonicTime
   -- TODO: revisit the policy once we have local root peers in the governor.
   -- The intention is to give local root peers give head start and avoid
   -- giving advantage to hostile and quick root peers.
   threadDelay 3
-  (mode, ubp) <- atomically ((,) <$> updateChurnMode
-                                 <*> getUseBootstrapPeers)
+  (_, regime) <- atomically $ pickChurnRegime <$> updateChurnMode
+                                              <*> getUseBootstrapPeers
+  -- peer selection governor starts in TooOld state, and unless bootstrap peers
+  -- are enabled, waits for initial peer targets set here, therefore to keep
+  -- things consistent, TooOld is used here to ensure that the right targets
+  -- are selected from the configuration.
+  let base = targetsSelector TooOld useGenesis
+  -- TODO: verify if #3396 is resolved
   atomically $ do
-    increaseActivePeers mode
-    increaseEstablishedPeers mode ubp
+    increaseActivePeers regime base
+    increaseEstablishedPeers regime base
+
+  -- Put back TooOld tag in the mutex to indicate to peer selection that we are done.
+  -- only the peer selection governor is responsible for switching this flag
+  -- on the basis of ledger state judgement to keep things organized.
+  atomically $ putTMVar mutexPeerSelection TooOld
   endTs0 <- getMonotonicTime
   fuzzyDelay inRng (endTs0 `diffTime` startTs0) >>= go
 
   where
 
+    pickChurnRegime :: ChurnMode -> UseBootstrapPeers -> (ChurnMode, ChurnRegime)
+    pickChurnRegime mode ubp =
+      (mode,) if | useGenesis -> Default
+                 | mode == ChurnModeBulkSync, UseBootstrapPeers _ <- ubp -> ChurnBootstrapSync
+                 | mode == ChurnModeBulkSync -> ChurnLegacySync
+                 | otherwise -> Default
```

Comment:
```suggestion
      ( mode
      , if | useGenesis
           -> Default
           | ChurnModeBulkSync <- mode
           , UseBootstrapPeers {} <- ubp
           -> ChurnBootstrapSync
           | ChurnModeBulkSync <- mode
           -> ChurnLegacySync
           | otherwise -> Default
      )
```
Actually, there's no reason to return `mode`. In the call site we are ignoring its value anyway.

Reply:
I kept it because there's a trace statement that logs this in the go loop.

Reply:
I see, there's another call site which doesn't ignore it.

In general pattern matches are nicer than `==`.